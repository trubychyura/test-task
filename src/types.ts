export type IComment = {
  text: string;
  rating: number;
};

export type IProduct = {
  asin: string;
  name: string;
  img: string;
  comments: IComment[];
  [key: string]: any;
};

export type HomeProps = {
  products: IProduct[];
  loadData: () => void;
};

export type RatingProps = {
  name?: string;
  stars?: number;
};

export type StarProps = {
  value: number;
  rating: number;
  hover?: number | null;
};

export type FormProps = {
  id: string;
  handleSubmit: (id: string, data: IComment) => void;
};

export type FormikValues = {
  comment: string;
  rating: number;
};

export type ProductProps = {
  id: string;
  name: string;
  img: string;
};

export type IState = {
  products: IProduct[];
};

export type ModalProps = {
  products: IProduct[];
};

export type ParamTypes = {
  id: string;
};
