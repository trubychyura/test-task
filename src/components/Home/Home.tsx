import React, { FC } from 'react';
import { connect } from 'react-redux';

import { Product } from '..';

import { loadData } from '../../redux/ducks/products';
import { IState, HomeProps } from '../../types';
import { StyledContainer, StyledTitle } from '../../styled/StyledHome';

export const Home: FC<HomeProps> = ({ products, loadData }) => {
  React.useEffect(() => {
    loadData();
  }, [loadData]);

  return (
    <StyledContainer>
      <StyledTitle>List of products</StyledTitle>
      {products.map(({ name, asin, img }) => (
        <Product id={asin} name={name} img={img} key={asin} />
      ))}
    </StyledContainer>
  );
};

export default connect(({ products }: IState) => ({ products }), { loadData })(
  Home,
);
