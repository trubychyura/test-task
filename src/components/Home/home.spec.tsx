import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { Home } from './Home';
import { Product } from '..';
import { HomeProps } from '../../types';

const setUp = (props: HomeProps): ShallowWrapper =>
  shallow(<Home {...props} />);

const products = [
  {
    asin: 'B0721KZHWT',
    name:
      'A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included',
    img:
      'https://images-na.ssl-images-amazon.com/images/I/71Y4Ogr8BNL._SX679_.jpg',
    comments: [
      {
        text: 'Some comment',
        rating: 4,
      },
    ],
  },
  {
    asin: 'B01HI4E470',
    name:
      'RoLightic 5M Waterproof Led Strip Light, Multi-Color,SMD 5050 150LEDs Color Changing Flexible Led Strip Kit with 44 key Remote Controller + Power Adapter for Home Kitchen Cabinet Car TV Lighting Decoration',
    img:
      'https://images-na.ssl-images-amazon.com/images/I/81GKt-J0kxL._SX425_.jpg',
    comments: [{ text: 'Some comment', rating: 3 }],
  },
  {
    asin: 'B078PG3PLC',
    name:
      'IPX7 Waterproof Bike Light Set-1000 Lumen Mini Bicycle Headlight Free Rear Light-USB Fast Charge High Performance Rechargeable Battery Easy Install & Universal Fit Mount',
    img:
      'https://images-na.ssl-images-amazon.com/images/I/81iRKp0s5JL._SX522_.jpg',
    comments: [{ text: 'Some comment', rating: 5 }],
  },
];

describe('Home component', () => {
  let component: ShallowWrapper;
  beforeEach(() => {
    const props = {
      products,
      loadData: () => {},
    };
    component = setUp(props);
  });

  it('should render component', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render defined amount of products', () => {
    const wrapper = component.find(Product);
    expect(wrapper).toHaveLength(products.length);
  });

  describe('useEffect hook', () => {
    let loadData: () => void;

    beforeEach(() => {
      jest.spyOn(React, 'useEffect').mockImplementation((f) => f());

      loadData = jest.fn().mockReturnValue(products);
      component = setUp({ products: [], loadData });
    });

    it('loads the products', () => {
      expect(loadData).toHaveBeenCalled();
    });
  });
});
