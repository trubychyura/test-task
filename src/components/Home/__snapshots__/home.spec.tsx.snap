// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should render component 1`] = `
<Styled(Container)>
  <styled.h1>
    List of products
  </styled.h1>
  <Product
    id="B0721KZHWT"
    img="https://images-na.ssl-images-amazon.com/images/I/71Y4Ogr8BNL._SX679_.jpg"
    key="B0721KZHWT"
    name="A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included"
  />
  <Product
    id="B01HI4E470"
    img="https://images-na.ssl-images-amazon.com/images/I/81GKt-J0kxL._SX425_.jpg"
    key="B01HI4E470"
    name="RoLightic 5M Waterproof Led Strip Light, Multi-Color,SMD 5050 150LEDs Color Changing Flexible Led Strip Kit with 44 key Remote Controller + Power Adapter for Home Kitchen Cabinet Car TV Lighting Decoration"
  />
  <Product
    id="B078PG3PLC"
    img="https://images-na.ssl-images-amazon.com/images/I/81iRKp0s5JL._SX522_.jpg"
    key="B078PG3PLC"
    name="IPX7 Waterproof Bike Light Set-1000 Lumen Mini Bicycle Headlight Free Rear Light-USB Fast Charge High Performance Rechargeable Battery Easy Install & Universal Fit Mount"
  />
</Styled(Container)>
`;
